File: TRViS.UITests/Pages/HorizontalTimetablePageObject.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Appium;
using OpenQA.Selenium.Support.UI;
using TRViS.UITests.Infrastructure;

namespace TRViS.UITests.Pages;

/// <summary>
/// Page object for the horizontal-timetable Shell route. The page is a single
/// AppBar + WebView (wrapped in a Grid that carries the AutomationId).
/// </summary>
public class HorizontalTimetablePageObject : PageObject
{
	public HorizontalTimetablePageObject(AppiumDriver driver) : base(driver) { }

	/// <summary>
	/// Finds the page's WebView. On iOS / macOS the wrapper Grid surfaces the
	/// MAUI AutomationId via the native accessibility identifier, so the
	/// standard AccessibilityId lookup is enough. Android and Windows are
	/// problematic in different ways: UIA2 does not forward AutomationId to
	/// <c>resource-id</c>, and WinUI 3 exposes the Grid as a non-control
	/// Pane that <c>AccessibilityId</c> cannot reach. On both we fall back
	/// to locating the sole native WebView node by its class name —
	/// <c>android.webkit.WebView</c> on Android and
	/// <c>Microsoft.UI.Xaml.Controls.WebView2</c> on Windows.
	/// </summary>
	public AppiumElement WebView => (IsAndroid || IsWindows)
		? WaitForWebViewByClassName()
		: WaitForElement(AutomationIds.HorizontalTimetable.WebView);

	private AppiumElement WaitForWebViewByClassName()
	{
		// On Windows we use XPath because Selenium .NET's By.ClassName maps to
		// the "css selector" strategy, which WinAppDriver rejects with
		// InvalidSelectorException. The Appium UIA2 (Android) server tolerates
		// it, so By.ClassName still works there.
		var wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(30));
		var locator = IsWindows
			? By.XPath("//*[@ClassName='Microsoft.UI.Xaml.Controls.WebView2']")
			: By.ClassName("android.webkit.WebView");
		Driver.Manage().Timeouts().ImplicitWait = TimeSpan.Zero;
		try
		{
			return (AppiumElement)wait.Until(d =>
			{
				try
				{
					var el = d.FindElement(locator);
					return el.Displayed ? el : null!;
				}
				catch (NoSuchElementException) { return null!; }
			});
		}
		finally
		{
			Driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
		}
	}

	public bool IsDisplayed()
	{
		try
		{
			return WebView.Displayed;
		}
		catch (NoSuchElementException)
		{
			return false;
		}
	}

	/// <summary>
	/// Taps the AppBar back button so the page pops back to DTAC. The
	/// HorizontalTimetable page is a Shell pushed page (not a flyout root),
	/// so the Shell flyout is not reachable here; getting back to a
	/// flyout-aware page is the prerequisite for the next fixture's
	/// <c>NavigateToHome</c> to work in shared-session runs. Returns the
	/// caller-friendly DTAC page object that the back navigation lands on.
	/// </summary>
	public DTACViewHostPageObject TapBack()
	{
		FindByAutomationId(AutomationIds.HorizontalTimetable.BackButton).Click();
		return new DTACViewHostPageObject(Driver);
	}
}

File: TRViS.UITests.Apple/Sources/Shared/HorizontalTimetablePageObject.swift
// HorizontalTimetablePageObject.swift
// Page object for the Horizontal Timetable page.
// Ports IsDisplayed() and TapBack() from
// TRViS.UITests/Pages/HorizontalTimetablePageObject.cs — Apple platform only.
//
// On iOS / macOS the wrapper Grid around the WebView carries
// AutomationId="HorizontalTimetable.WebView" (see HorizontalTimetablePage.cs
// comments), so a plain AccessibilityIdentifier lookup is sufficient.
// The Android/Windows class-name fallbacks in the C# port are not needed here.

import XCTest

class HorizontalTimetablePageObject {

    private let app: XCUIApplication
    private let base: BaseUITestCase

    init(app: XCUIApplication, base: BaseUITestCase) {
        self.app = app
        self.base = base
    }

    // MARK: — Display check (ported from C# IsDisplayed)

    /// True when the HorizontalTimetable page's WebView wrapper Grid is in the
    /// accessibility tree and has a non-zero frame (i.e., it is laid out).
    func isDisplayed(timeout: TimeInterval = 30) -> Bool {
        guard let el = base.waitForElement(
            id: AutomationIds.HorizontalTimetable.webView, timeout: timeout
        ) else {
            return false
        }
        return el.exists && el.frame.size.width > 0 && el.frame.size.height > 0
    }

    // MARK: — Navigation (ported from C# TapBack)

    /// Taps the AppBar back button to pop back to DTAC.
    /// HorizontalTimetable is a Shell.GoToAsync pushed page (not a flyout root),
    /// so the Shell flyout is not reachable from here. Returning to DTAC
    /// ensures the next fixture's NavigateToHome via the flyout can work.
    @discardableResult
    func tapBack() -> DTACViewHostPageObject {
        guard let btn = base.waitForElement(
            id: AutomationIds.HorizontalTimetable.backButton, timeout: 15
        ) else {
            XCTFail("HorizontalTimetable.BackButton not found")
            return DTACViewHostPageObject(app: app, base: base)
        }
        btn.tap()
        return DTACViewHostPageObject(app: app, base: base)
    }
}
